h/services/oauth.py

```diff
@@ -161,10 +144,51 @@ def create_token(self, user, authclient):
 
         return token
 
-    def _decode(self, token, **kwargs):
+
+class GrantToken(object):
+    """
+    Represents a JWT bearer grant token.
+
+    This class is responsible for a couple of things: firstly, verifying that
+    the token is a correctly-formatted JSON Web Token, and that it contains all
+    the required claims in the right formats. Some of this processing is
+    deferred to the `jwt` module, but that doesn't handle all the fields we
+    want to validate.
+
+    """
+
+    def __init__(self, token):
+        if not token or type(token) != text_type:
+            raise OAuthTokenError('required assertion parameter is missing',
+                                  'invalid_request')
+        self._token = token
+
+        try:
+            self._claims = jwt.decode(token, verify=False)
+        except jwt.DecodeError:
+            raise OAuthTokenError('invalid JWT signature', 'invalid_grant')
+
+    @property
+    def issuer(self):
+        iss = self._claims.get('iss', None)
+        if not iss:
+            raise OAuthTokenError('grant token issuer is missing', 'invalid_grant')
+        return iss
+
+    @property
+    def subject(self):
```

Comment:
One thing I'm missing is that the old code was very clear that it first decodes the claims without verifying for getting the issuer, and then decoded the claims again with verification for reading the other claims (by using `unverified_claims` and `claims` variable names).
Since this isn't a general purpose grant token class and is only being used here, we might want to track the verification state and raise an error if we try to access the subject (or any other non-issuer) claim without first verifying the grant token?

What do you think?